File: crates/idakit-sys-codegen/src/domains/strings.rs
use super::super::model::*;

/// The strings domain: IDA's string list plus per-literal decoding. `strlist_build` runs an
/// O(database) scan to (re)build the list; `strlist_item` returns the nth entry as a `StrlistItem`
/// (throws when out of range). `strlit_contents` decodes one literal semantically (`STRCONV_REPLCHAR`,
/// undecodable units to U+FFFD) and `strlit_escaped` to its display form (`STRCONV_ESCAPE`, C-escaped
/// as in the pseudocode); both throw only when the literal cannot be read. All bodies are
/// hand-written in `facade/strings.cpp`.
pub const STRINGS: Domain = Domain {
    name: "strings",
    sdk_includes: &["<strlist.hpp>", "<bytes.hpp>", "<stdexcept>"],
    externs: &[],
    structs: &[SharedStruct {
        name: "StrlistItem",
        doc: "One string-list entry: its address, octet length, and `STRTYPE` code.",
        fields: fields! {
            ea: U64 = "Address of the string literal.";
            length: I32 = "Length in octets (raw bytes, not decoded characters).";
            type_: I32 = "`STRTYPE` code describing the encoding.";
        },
    }],
    consts: &[],
    custom_tus: &["facade/strings.cpp"],
    fns: fns! {
        "(Re)build IDA's string list, an O(database) scan of the whole image."
            strlist_build();
        "Number of entries in the current string list (`get_strlist_qty`)."
            strlist_qty() -> Usize;
        "The `n`-th string-list entry as a `StrlistItem`; `Err` when `n` is out of range."
            strlist_item(n: Usize) -> ResultShared("StrlistItem");
        "Decode the string literal at `ea` (given octet length and `STRTYPE`) semantically \
         (`STRCONV_REPLCHAR`: undecodable units become U+FFFD); `Err` when it cannot be read."
            strlit_contents(ea: U64, len: Usize, strtype: I32) -> ResultString;
        "Decode the string literal at `ea` to its C-escaped display form (`STRCONV_ESCAPE`, as the \
         decompiler renders it); `Err` when it cannot be read."
            strlit_escaped(ea: U64, len: Usize, strtype: I32) -> ResultString;
    },
};

File: crates/idakit-sys/facade/strings.cpp
// Hand-written Custom bodies for the generated strings domain (namespace gen). Two independent
// pieces: IDA's cached string list (build/query/index into StrlistItem, a cxx shared struct
// defined by the cxx-generated gen_bridge.h), and strlit_contents/strlit_escaped, which decode a
// string literal at a given address directly, list or not. Every out-of-range index or unreadable
// literal throws, surfacing as a Rust Err; STRCONV_REPLCHAR/STRCONV_ESCAPE both guarantee the
// decoded bytes are valid UTF-8.

#include <ida.hpp>
#include <pro.h>

#include <bytes.hpp>   // get_strlit_contents, STRCONV_REPLCHAR, STRCONV_ESCAPE
#include <strlist.hpp> // build_strlist, get_strlist_qty, get_strlist_item, string_info_t

#include <stdexcept>

#include "gen_strings.h"
// The cxx-generated header defines the StrlistItem shared struct (full definition needed to
// construct it); gen_strings.h only forward-declares it.
#include "gen_bridge.h"

namespace gen {

// Rebuilds IDA's cached string list; an O(database) scan.
void strlist_build() { build_strlist(); }

// The number of entries in the cached string list.
size_t strlist_qty() { return get_strlist_qty(); }

// The n-th entry of the cached string list; throws if n is out of range.
StrlistItem strlist_item(size_t n) {
  string_info_t info;
  if (!get_strlist_item(&info, n))
    throw std::out_of_range("string list index out of range");
  StrlistItem item;
  item.ea = static_cast<uint64_t>(info.ea);
  item.length = static_cast<int32_t>(info.length);
  item.type_ = static_cast<int32_t>(info.type);
  return item;
}

// The string literal at addr, decoded with undecodable bytes replaced by U+FFFD; throws if the
// literal can't be read.
rust::String strlit_contents(uint64_t addr, size_t len, int32_t strtype) {
  qstring out;
  ssize_t r =
      get_strlit_contents(&out, static_cast<ea_t>(addr), len, strtype, nullptr, STRCONV_REPLCHAR);
  if (r < 0)
    throw std::runtime_error("unreadable string literal");
  return to_rust_string(out);
}

// The string literal at addr, decoded to its C-escaped display form (\n, \xNN, ...); throws if
// the literal can't be read.
rust::String strlit_escaped(uint64_t addr, size_t len, int32_t strtype) {
  qstring out;
  ssize_t r =
      get_strlit_contents(&out, static_cast<ea_t>(addr), len, strtype, nullptr, STRCONV_ESCAPE);
  if (r < 0)
    throw std::runtime_error("unreadable string literal");
  return to_rust_string(out);
}

} // namespace gen
